### DISCUSSION TITLE
How to associate file metadata with file in storage

### DISCUSSION THREAD

**Discussion by greensomes** [directus/directus#17458] [Archive]
Hello! I'm attempting to associate metadata with a file already in AWS storage by following guidance [in the docs](https://docs.directus.io/reference/files.html#upload-a-file).

As stated: 

_Alternatively, you can use application/json with JSON request body to associate metadata to a file that already exists in the storage. The type property will be required to specify the mimetype of that file._

I've attempted a few things here, including a POST to /files with the file object as JSON payload. In this case, I receive the following response:

```
{
    "errors": [
        {
            "message": "Unsupported Content-Type header",
            "extensions": {
                "code": "UNSUPPORTED_MEDIA_TYPE"
            }
        }
    ]
}
```

The Content-Type header in this case is application/json. It seems obvious why this approach fails, as the endpoint does not accept application/json.

Can someone provide an example of how updating metadata for a file existing in storage is done?

Many thanks in advance.

**Comment by paescuj** [comment]
If you want to add metadata to an already existing file, you should use the update (`PATCH`) method: https://docs.directus.io/reference/files.html#update-a-file

**Reply by greensomes** [reply]
The file doesn't exist in directus_files yet, so that won't work. I was hoping that there was a way to insert into directus_files while referencing a file already in storage. 

Some background on what we're trying to do may be helpful. We have two instances of Directus running: preview and prod. We're using flows to publish approved content from preview to prod. Since these instances are backed by the same s3 bucket, it was my hope to just push the file metadata from one directus instance to another while pointing to the same file that resides in s3. 

I may have to settle for a slightly less desirable solution that duplicates files in s3 to achieve my goal.

**Reply by paescuj** [reply]
I see, thanks for the clarification! One possibility might be to export the files collection on the preview instance using the [export API endpoint](https://docs.directus.io/reference/system/utilities.html#export-data-to-a-file) and then import it on the prod instance with the help of the [import API endpoint](https://docs.directus.io/reference/system/utilities.html#import-data-from-file).

**Reply by greensomes** [reply]
I like where this is going. The export worked like a charm, but I'm having issues with import. I'm attempting a proof of concept via Postman and the response I receive is:

```
{
    "errors": [
        {
            "message": "You don't have permission to access this.",
            "extensions": {
                "code": "FORBIDDEN"
            }
        }
    ]
}
```

Here's the code snippet generated by Postman:

```
POST /utils/import/directus_files HTTP/1.1
Host: my-host
Authentication: Bearer my-token
Content-Length: 213
Content-Type: multipart/form-data; boundary=----WebKitFormBoundary7MA4YWxkTrZu0gW

----WebKitFormBoundary7MA4YWxkTrZu0gW
Content-Disposition: form-data; name="file"; filename="export-directus_files-20230209-16637.json"
Content-Type: application/json

(data)
----WebKitFormBoundary7MA4YWxkTrZu0gW
```

I'm still tinkering, but let me know if anything jumps out at you.

Thanks!

**Reply by paescuj** [reply]
Hmm, I assume you're uploading with the admin user, right? Do you see additional information about the error on the server console?

**Reply by greensomes** [reply]
Yes, I've tried using bearer tokens for a service account user and an admin user as well as sending requests without an Authorization header. All get the same result.

I looked at the logs and I see the requests come in, but there are no errors or otherwise meaningful messages there.

**Reply by paescuj** [reply]
Okay and does it work if you upload the export via the app? If so you might be able to compare the requests.

**Reply by greensomes** [reply]
Yes, the import works from the app. The app logs don't reveal anything other than the obvious. And inspecting the headers of the request from the app didn't seem to have anything missing from the Postman request

Out of curiosity, I checked to see if the upload file API via Postman gave me the same unauthorized error as the import. The file upload completed successfully.

**Reply by greensomes** [reply]
I also tried the import via a script operation in a flow:
```
var FormData = require('form-data');
const axios = require('axios');

module.exports = async function(data) {

    try {
        const exportFile = await axios.get('https://my-preview-directus-instance/assets/437b2ff0-74c5-4032-841d-a14ea24aa7db', { responseType: 'stream' });

        const form = new FormData();
        form.append('file', exportFile.data, 'export.json');

        const response = await axios.post('https://my-prod-directus-instance/utils/import/directus_files', form, {
          headers: {
            ...form.getHeaders(),
              Authorization: 'Bearer admin-account-token'
          },
        });
    
	return response;
    }
    catch (e) {
        return { "exception": e.message};
    }
}
```

The response from the above is
```
{
  "exception": "Request failed with status code 403"
}
```

**Reply by greensomes** [reply]
The file upload works fine:
```
var FormData = require('form-data');
const axios = require('axios');

module.exports = async function(data) {

    try {
        const fileResponse = await axios.get('https://upload.wikimedia.org/wikipedia/en/d/df/The_Odd_Squad.png', { responseType: 'stream' });

        const form = new FormData();
        form.append('file', fileResponse.data, 'odd-squad.png');

        const response = await axios.post('https://my-directus-instance/files', form, {
          headers: {
            ...form.getHeaders()
          },
        });
    
	return { "status": response.status };
    }
    catch (e) {
        return { "exception": e.message};
    }
}
```

Because the [documentation](https://docs.directus.io/reference/system/utilities.html#import-data-from-file) states that this behaves similar to a file upload, this is feeling like a bug. The documentation doesn't say anything about settings or user privileges specific to this operation. Thoughts?

**Reply by paescuj** [reply]
> Yes, the import works from the app. The app logs don't reveal anything other than the obvious. And inspecting the headers of the request from the app didn't seem to have anything missing from the Postman request
> 
> Out of curiosity, I checked to see if the upload file API via Postman gave me the same unauthorized error as the import. The file upload completed successfully.

IMO if it works via the app, you should be able to do the same manually (AFAIK the app uses the API for this as well). Maybe try it without Postman...

**Reply by greensomes** [reply]
Thanks so much for your assistance on this. I managed to get the import to work as expected via the API in flows. I made the mistake of coding tired and wasn't properly setting the Authorization header in the request.

Many thanks!

**Reply by paescuj** [reply]
Oh that's great! 😃 You're welcome! I'm very glad it's working now 👍
